src/metaswitch/clearwater/config_manager/etcd_synchronizer.py

```diff
@@ -154,3 +122,50 @@ def read_from_etcd(self):
             # function again after we return, causing the read to be retried.
 
         return value
+
+    def waitHelper(self, timeout, waitIndex):
+        result = None
+        # Calculate the args for the wait
+        args = dict(wait=True,
+                    timeout=timeout,
+                    recursive=False,
+                    quorum=True)
+        if waitIndex is not None:
+            args['waitIndex'] = result.modifiedIndex + 1
+
+        try:
+            # Wait for the key to change
+            result = self._client.read(full_key,
+                                       **args)
+            break
+        except urllib3.exceptions.TimeoutError:
+            # Timeouts after 5 seconds are expected, so ignore them
+            # - unless we're terminating, we'll stay in the while
+            # loop and try again
+            pass
+        except etcd.EtcdEventIndexCleared as e:
+            # A snapshot has occurred since the last change, invalidating the
+            # index, so wait again without specifying an index.
+            # This introduces a race condition (a change could happen between
+            # the two watches which we wouldn't spot), so set a 10-minute time
+            # limit on the watch as a safeguard.
+            _log.info("An etcd snapshot has invalidated index %d - "
+                      "starting a new watch (timeout 600s) without specifying an index",
+                      args.get('waitIndex', None))
+            result = self.waitHelper(timeout=600, waitIndex=None)
+        except etcd.EtcdException as e:
+            # We have seen timeouts getting raised as EtcdExceptions
+            # so catch these here too and treat them as timeouts if
+            # they indicate that the read timed out.
+            if "Read timed out" in e.message:
+                pass
+            else:
+                raise
+        except ValueError:
+            # The index isn't valid to watch on, probably because
+            # there has been a snapshot between the get and the
+            # watch. Just start the read again.
```

Comment:
This comment is now bogus, and invalidated index goes through the new catch branch.